internal/signer/linux/pkcs11/pkcs11.go

```diff
@@ -139,3 +145,52 @@ func (k *Key) Public() crypto.PublicKey {
 func (k *Key) Sign(_ io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
 	return k.signer.Sign(nil, digest, opts)
 }
+
+func (k *Key) Encrypt(data []byte) ([]byte, error) {
+	publicKey := k.Public()
+	_, ok := publicKey.(*rsa.PublicKey)
+	if ok {
+		return k.encryptRSAGoPKCS11(data)
+	}
+	_, ok = publicKey.(*ecdsa.PublicKey)
+	if ok {
+		// TODO: implement encryption for ec keys
+		return nil, nil
+	}
+	return nil, fmt.Errorf("encrypt error: Unsupported key type")
+}
+
+func (k *Key) Decrypt(encryptedData []byte) ([]byte, error) {
+	publicKey := k.Public()
+	_, ok := publicKey.(*rsa.PublicKey)
+	if ok {
+		return k.decryptRSAGoPKCS11(encryptedData)
+	}
+	_, ok = publicKey.(*ecdsa.PublicKey)
+	if ok {
+		// TODO: implement decryption for ec keys
+		return nil, nil
+	}
+	return nil, fmt.Errorf("decrypt error: Unsupported key type")
+}
+
+func (k *Key) encryptRSA(hash hash.Hash, data []byte) ([]byte, error) {
+	publicKey := k.Public()
+	rsaPubKey := publicKey.(*rsa.PublicKey)
+	return rsa.EncryptOAEP(hash, rand.Reader, rsaPubKey, data, nil)
+}
+
+func (k *Key) encryptRSAGoPKCS11(data []byte) ([]byte, error) {
+	publicKeyFilter := pkcs11.Filter{pkcs11.ClassPublicKey, ""}
```

Comment:
What is going on here? Does it work?

Seem like it just grabs all PublicKeys and grabs the first one? Don't we have a label in the ECP config we could use?

Reply:
It searches the slot for public keys and returns the first one. It does work when getting _a_ public key. I assumed that there was only one matching public key when creating the filter, but there may be cases with more than one. I'm not sure which label would apply or where to find it though

Reply:
[Here](https://github.com/googleapis/enterprise-certificate-proxy/blob/main/internal/signer/linux/signer.go#L112) is where the PKCS#11 cred object is initialized using a label. Assume that the public key, private key, and certificate all share the same label.

This should be enough to find the correct public key.

Reply:
Ah okay, I will use that. Thank you

Reply:
I added a label field to the key struct and used that. Is that okay, or should I take a different approach?

Reply:
That looks good!